File: solutions/java/satellite/1/src/main/java/Satellite.java
import java.util.List;
import java.util.HashSet;
import java.util.Set;
import java.util.HashMap;
import java.util.Map;

public class Satellite {
    
    public Tree treeFromTraversals(List<Character> preorderInput, List<Character> inorderInput) {
        // 1. Validación de longitudes
        if (preorderInput.size() != inorderInput.size()) {
            throw new IllegalArgumentException("traversals must have the same length");
        }
        
        // 2. Validación de elementos únicos
        Set<Character> preSet = new HashSet<>(preorderInput);
        Set<Character> inSet = new HashSet<>(inorderInput);
        
        if (preSet.size() != preorderInput.size() || inSet.size() != inorderInput.size()) {
            throw new IllegalArgumentException("traversals must contain unique items");
        }
        
        // 3. Validación de que contienen los mismos elementos
        if (!preSet.equals(inSet)) {
            throw new IllegalArgumentException("traversals must have the same elements");
        }
        
        // Optimización: Mapa para encontrar el índice de cualquier nodo en el recorrido in-order en O(1)
        Map<Character, Integer> inOrderIndex = new HashMap<>();
        for (int i = 0; i < inorderInput.size(); i++) {
            inOrderIndex.put(inorderInput.get(i), i);
        }
        
        // 4. Construcción recursiva del árbol
        Node root = buildTree(preorderInput, 0, preorderInput.size() - 1,
                              inorderInput, 0, inorderInput.size() - 1, 
                              inOrderIndex);
                              
        return new Tree(root);
    }

    private Node buildTree(List<Character> preorder, int preStart, int preEnd,
                           List<Character> inorder, int inStart, int inEnd,
                           Map<Character, Integer> inOrderIndex) {
        
        // Caso base: no hay nodos para procesar en este subárbol
        if (preStart > preEnd) {
            return null;
        }
        
        // El primer elemento del pre-order siempre es la raíz del subárbol actual
        char rootVal = preorder.get(preStart);
        Node root = new Node(rootVal);
        
        // Si solo hay un nodo, es una hoja
        if (preStart == preEnd) {
            return root;
        }
        
        // Encontramos la raíz en el recorrido in-order para dividir los subárboles
        int rootIdx = inOrderIndex.get(rootVal);
        int leftSize = rootIdx - inStart; // Cantidad de nodos en el subárbol izquierdo
        
        // Construir recursivamente el subárbol izquierdo
        root.left = buildTree(preorder, preStart + 1, preStart + leftSize,
                              inorder, inStart, rootIdx - 1, 
                              inOrderIndex);
                              
        // Construir recursivamente el subárbol derecho
        root.right = buildTree(preorder, preStart + leftSize + 1, preEnd,
                               inorder, rootIdx + 1, inEnd, 
                               inOrderIndex);
                               
        return root;
    }
}

File: solutions/javascript/satellite/1/satellite.js
//
// This is only a SKELETON file for the 'Satellite' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const treeFromTraversals = (preorder, inorder) => {
  // Validaciones
  if (preorder.length !== inorder.length) {
    throw new Error('traversals must have the same length');
  }
  
  if (new Set(preorder).size !== preorder.length || 
      new Set(inorder).size !== inorder.length) {
    throw new Error('traversals must contain unique items');
  }
  
  if (new Set(preorder).size !== new Set(inorder).size ||
      !preorder.every(item => inorder.includes(item))) {
    throw new Error('traversals must have the same elements');
  }
  
  // Función recursiva para construir el árbol
  function build(pre, ino) {
    // Caso base: si no hay elementos, devolver objeto vacío
    if (pre.length === 0) return {};
    
    // El primer elemento del pre-order es la raíz
    const rootValue = pre[0];
    const rootIndex = ino.indexOf(rootValue);
    
    // Dividir el in-order en subárbol izquierdo y derecho
    const leftInorder = ino.slice(0, rootIndex);
    const rightInorder = ino.slice(rootIndex + 1);
    
    // Dividir el pre-order en consecuencia
    // Los siguientes N elementos (tamaño del subárbol izquierdo) son del izquierdo
    const leftPreorder = pre.slice(1, 1 + leftInorder.length);
    const rightPreorder = pre.slice(1 + leftInorder.length);
    
    // Construir recursivamente el nodo actual
    return {
      value: rootValue,
      left: build(leftPreorder, leftInorder),
      right: build(rightPreorder, rightInorder)
    };
  }
  
  return build(preorder, inorder);
};